### PR TITLE
fix(compiler): allow recursive objects to be used as DI values

### PR DESCRIPTION
**Pull request by mattlewis92**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] <s>Docs have been added / updated (for bug fixes / features)</s>

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
https://github.com/angular/angular/issues/13685

**What is the new behavior?**
The compiler no longer throws when trying to use a recursive object as a dependency injection value. 

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: modules/@angular/compiler/src/util.ts [hunk 1]
```diff
@@ -35,40 +35,52 @@ function _splitAt(input: string, character: string, defaultValues: string[]): st
   return [input.slice(0, characterIndex).trim(), input.slice(characterIndex + 1).trim()];
 }
 
-export function visitValue(value: any, visitor: ValueVisitor, context: any): any {
+export function visitValue(
```
Comment (vicb) [inline modules/@angular/compiler/src/util.ts]: could `visited` be (part of) the context ?

File: modules/@angular/compiler/src/util.ts [hunk 2]
```diff
@@ -35,40 +35,52 @@ function _splitAt(input: string, character: string, defaultValues: string[]): st
...
+  visitStringMap(map: {[key: string]: any}, context: any, visited: WeakMap<any, any>): any;
+  visitPrimitive(value: any, context: any, visited: WeakMap<any, any>): any;
+  visitOther(value: any, context: any, visited: WeakMap<any, any>): any;
 }
 
 export class ValueTransformer implements ValueVisitor {
-  visitArray(arr: any[], context: any): any {
-    return arr.map(value => visitValue(value, this, context));
+  visitArray(arr: any[], context: any, visited: WeakMap<any, any> = new WeakMap()): any {
+    return arr.map(value => visitValue(value, this, context, visited));
```
Comment (vicb) [inline modules/@angular/compiler/src/util.ts]: handle recursion for `Array` ?

File: modules/@angular/compiler/src/util.ts (hunk 1 above)
Comment (mattlewis92) [inline modules/@angular/compiler/src/util.ts]: As the context is passed by the user of the function there's no guarantee context will be an object i.e. context could be a string or an array etc

File: modules/@angular/compiler/src/util.ts (hunk 2 above)
Comment (mattlewis92) [inline modules/@angular/compiler/src/util.ts]: Done

File: modules/@angular/compiler/src/util.ts (hunk 1 above)
Comment (mattlewis92) [inline modules/@angular/compiler/src/util.ts]: e.g. [here the context is null](https://github.com/angular/angular/blob/master/modules/%40angular/compiler/src/output/value_util.ts#L23) so you wouldn't be able to attach `visited` to it. In other places context is the calling class instance so you'd then end up polluting it with the visited weakmap (you could mitigate this by using a symbol as the object key and removing it once traversal is done, but that just feels really messy)

**Comment by mattlewis92** [comment]
@vicb does anything else need to be done for this to be merged? 😄

**Comment by vicb** [comment]
It needs to be reviewed by the owner of the component.

**Comment by mattlewis92** [comment]
OK, thanks for letting me know!

**Comment by tbosch** [comment]
In 5.0 we plan on supporting arbitrary values in `useValue` providers, so this change is obsolete.

**Pull request closed without merging**